synapse/cortex.py

```diff
@@ -1379,6 +1381,72 @@ async def getDeprLocks(self):
 
         return retn
 
+    async def reqValidStormGraph(self, gdef):
+        for filt in gdef.get('filters', ()):
+            await self.getStormQuery(filt)
+
+        for pivo in gdef.get('filters', ()):
+            await self.getStormQuery(pivo)
+
+        for form, rule in gdef.get('forms', {}).items():
+            if form != '*' and self.model.form(form) is None:
+                raise s_exc.NoSuchForm(name=form)
+
+            for filt in rule.get('filters', ()):
+                await self.getStormQuery(filt)
+
+            for pivo in rule.get('filters', ()):
+                await self.getStormQuery(pivo)
+
+    async def addStormGraph(self, gdef):
+        gdef['iden'] = s_common.guid()
+
+        s_stormlib_graph.reqValidGdef(gdef)
+
+        await self.reqValidStormGraph(gdef)
+
+        return await self._push('storm:graph:add', gdef)
+
+    @s_nexus.Pusher.onPush('storm:graph:add')
+    async def _addStormGraph(self, gdef):
+        s_stormlib_graph.reqValidGdef(gdef)
+
+        await self.reqValidStormGraph(gdef)
+
+        iden = gdef['iden']
+        if self.graphs.get(iden) is not None:
+            return
+
+        self.graphs.set(iden, gdef)
+        await self.feedBeholder('storm:graph:add', {'gdef': gdef})
+        return copy.deepcopy(gdef)
+
+    async def delStormGraph(self, iden):
+        if self.graphs.get(iden) is None:
+            mesg = f'No graph projection with iden {iden} exists!'
+            raise s_exc.NoSuchIden(mesg=mesg)
+
+        return await self._push('storm:graph:del', iden)
+
+    @s_nexus.Pusher.onPush('storm:graph:del')
+    async def _delStormGraph(self, iden):
+        gdef = self.graphs.get(iden)
+        if gdef is None:
+            return
+
+        self.graphs.pop(iden)
+        await self.feedBeholder('storm:graph:del', {'iden': iden})
+
+    async def getStormGraph(self, iden):
+        gdef = self.graphs.get(iden)
+        if gdef is None:
+            mesg = f'No graph projection exists with iden {iden}.'
+            raise s_exc.NoSuchIden(mesg=mesg)
+        return copy.deepcopy(gdef)
+
+    async def getStormGraphs(self):
+        return [copy.deepcopy(g[1]) for g in self.graphs.items()]
```

Comment:
is it possible to stream this as opposed to returning a ( potentially large ) list ?

Reply:
That may require changing the SlabDict behavior slightly. We'll need to discuss that.

Reply:
changed to a generator